rasterio/merge.py

```diff
@@ -396,74 +400,90 @@ def nullcontext(obj):
         else:
             chunks = [dout_window]
 
-        logger.debug("Chunks=%r", chunks)
+        def _intersect_bounds(bounds1, bounds2, transform):
+            """Based on gdal_merge.py."""
+            int_w = max(bounds1[0], bounds2[0])
+            int_e = min(bounds1[2], bounds2[2])
+
+            if int_w >= int_e:
+                raise ValueError
+
+            if transform.e < 0:
+                # north up
+                int_s = max(bounds1[1], bounds2[1])
+                int_n = min(bounds1[3], bounds2[3])
+                if int_s >= int_n:
+                    raise ValueError
+            else:
+                int_s = min(bounds1[1], bounds2[1])
+                int_n = max(bounds1[3], bounds2[3])
+                if int_n >= int_s:
+                    raise ValueError
+
+            return int_w, int_s, int_e, int_n
 
         for chunk in chunks:
             dst_w, dst_s, dst_e, dst_n = windows.bounds(chunk, output_transform)
             dest = np.zeros((output_count, chunk.height, chunk.width), dtype=dt)
             if inrange:
                 dest.fill(nodataval)
 
+            # From gh-2221
+            chunk_bounds = windows.bounds(chunk, output_transform)
+            chunk_transform = windows.transform(chunk, output_transform)
+
             for idx, dataset in enumerate(sources):
                 with dataset_opener(dataset) as src:
-                    # 0. Precondition checks
-                    #    - Check that source is within destination bounds
-                    #    - Check that CRS is same
-
-                    if disjoint_bounds((dst_w, dst_s, dst_e, dst_n), src.bounds):
-                        logger.debug(
-                            "Skipping source: src=%r, bounds=%r",
-                            src,
-                            (dst_w, dst_s, dst_e, dst_n),
-                        )
-                        continue
 
+                    # Intersect source bounds and tile bounds
                     if first_crs != src.crs:
                         raise RasterioError(f"CRS mismatch with source: {dataset}")
 
-                    # 1. Compute the source window
-                    src_window = windows.from_bounds(
-                        dst_w, dst_s, dst_e, dst_n, src.transform
-                    ).round(3)
+                    try:
+                        ibounds = _intersect_bounds(
+                            src.bounds, chunk_bounds, chunk_transform
+                        )
+                        sw = windows.from_bounds(*ibounds, src.transform)
+                        cw = windows.from_bounds(*ibounds, chunk_transform)
+                    except (ValueError, WindowError):
+                        logger.info(
+                            "Skipping source: src=%r, bounds=%r", src, src.bounds
+                        )
+                        continue
 
-                    temp_shape = (src_count, chunk.height, chunk.width)
+                    cw = cw.align()
```

Comment:
New! This does the kind of aligning in `gdal_merge.py`.

Reply:
I like the new method. I noticed in `.align()`, 0.1 is added to the offsets, however, 0.001 is added to the offsets in `.round_offsets()`.